homeassistant/components/efesto/config_flow.py

```diff
@@ -0,0 +1,104 @@
+"""Config flow for Efesto."""
+from collections import OrderedDict
+import logging
+
+from efestoclient import (  # pylint: disable=redefined-builtin
+    ConnectionError,
+    EfestoClient,
+    Error as EfestoError,
+    InvalidURLError,
+    UnauthorizedError,
+)
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import (
+    CONF_DEVICE,
+    CONF_NAME,
+    CONF_PASSWORD,
+    CONF_URL,
+    CONF_USERNAME,
+)
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def device_entries(hass):
+    """Return the host,port tuples for the domain."""
+    return set(
+        entry.data[CONF_DEVICE] for entry in hass.config_entries.async_entries(DOMAIN)
+    )
+
+
+class EfestoConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Efesto Config Flow handler."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    def _entry_in_configuration_exists(self, user_input) -> bool:
+        """Return True if device already exists in configuration."""
+        device_id = user_input[CONF_DEVICE]
+        if device_id in device_entries(self.hass):
+            return True
+        return False
+
+    async def async_step_user(self, user_input=None):
+        """User initiated integration."""
+        errors = {}
+        if user_input is not None:
+            # Validate user input
+            url = user_input[CONF_URL]
+            username = user_input[CONF_USERNAME]
+            password = user_input[CONF_PASSWORD]
+            device_id = user_input[CONF_DEVICE]
+            name = user_input.get(CONF_NAME, device_id)
+
+            if self._entry_in_configuration_exists(user_input):
+                return self.async_abort(reason="device_already_configured")
+
+            try:
+                client = EfestoClient(url, username, password, device_id)
+                client.get_status()
+            except UnauthorizedError:
+                errors["base"] = "unauthorized"
+            except ConnectionError:
+                errors["base"] = "connection_error"
+            except InvalidURLError:
+                errors["base"] = "invalid_url"
+            except EfestoError:
+                errors["base"] = "unknown_error"
+
+            if "base" not in errors:
+                return self.async_create_entry(
+                    title=name,
+                    data={
+                        CONF_URL: url,
+                        CONF_USERNAME: username,
+                        CONF_PASSWORD: password,
+                        CONF_DEVICE: device_id,
+                    },
+                )
+        else:
+            user_input = {}
+
+        data_schema = OrderedDict()
```

Comment:
Please use a schema directly instead of the `OrderedDict()`, e.g.
https://github.com/home-assistant/home-assistant/blob/6f8f23238ad2db3acfc949cb497dca33dd604ab4/homeassistant/components/abode/config_flow.py#L28-L31